File: mvvm/JetLiYu/app/src/main/java/pri/weiqiang/jet/liyujapanese/ui/word/WordFragment.java
package pri.weiqiang.jet.liyujapanese.ui.word;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import pri.weiqiang.jet.liyujapanese.databinding.FragmentWordBinding;
import pri.weiqiang.jet.liyujapanese.ui.adapter.WordsRecyclerAdapter;

public class WordFragment extends Fragment {
    private String TAG = WordFragment.class.getSimpleName();
    private FragmentWordBinding binding;
    private WordFragmentViewModel model;
    private RecyclerView rvWord;
    private WordsRecyclerAdapter adapter;
    private boolean isExpandable = false;
    @Override
    public View onCreateView(
            LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState
    ) {

        binding = FragmentWordBinding.inflate(inflater, container, false);
        return binding.getRoot();

    }

    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        binding.buttonFirst.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
//                NavHostFragment.findNavController(WordFragment.this).navigate(R.id.action_FirstFragment_to_SecondFragment);
                model.setQuery("新标日初级_02");
//                model.getWordList("新标日初级_01").observe(getViewLifecycleOwner(),words->{
//                    if (words != null && words.size() != 0) {
//                        Log.e(TAG, "words.get(0):" + words.get(0).toString());
//                    }
//                });
            }
        });
        binding.btnGet.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                model.setQuery("新标日初级_03");
//                model.getWordList("新标日初级_01").observe(getViewLifecycleOwner(),words->{
//                    if (words != null && words.size() != 0) {
//                        Log.e(TAG, "words.get(0):" + words.get(0).toString());
//                    }
//                });
            }
        });

        model = new ViewModelProvider(this).get(WordFragmentViewModel.class);
        //ViewModel 绝不能引用视图、Lifecycle 或可能存储对 Activity 上下文的引用的任何类。
//        model.setContext(getActivity());
        model.getWordList().observe(getViewLifecycleOwner(), words -> {
            if (words != null && words.size() != 0) {
                Log.e(TAG, "words:"+words+" words.get(0):" + words.get(0).toString());
            }
            if (adapter==null){
                adapter = new WordsRecyclerAdapter(getActivity(),words,isExpandable);
                LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
                linearLayoutManager.setOrientation(RecyclerView.VERTICAL);
                binding.rvWord.setLayoutManager(linearLayoutManager);
                binding.rvWord.setAdapter(adapter);
                adapter.notifyDataSetChanged();
            }else {
                adapter.updateWordList(words);
                adapter.notifyDataSetChanged();
            }

        });
        Log.e(TAG, "wordList:" + model.getWordList());
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

}

File: mvvm/KtJetLiYu/app/src/main/java/pri/weiqiang/jet/kt/liyujapanese/ui/fragment/WordFragment.kt
package pri.weiqiang.jet.kt.liyujapanese.ui.fragment

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import pri.weiqiang.jet.kt.liyujapanese.R
import pri.weiqiang.jet.kt.liyujapanese.data.bean.Word
import pri.weiqiang.jet.kt.liyujapanese.databinding.FragmentWordBinding
import pri.weiqiang.jet.kt.liyujapanese.ui.adapter.WordAdapter
import pri.weiqiang.jet.kt.liyujapanese.ui.viewmodel.WordFragmentViewModel

/**
 * A simple [Fragment] subclass as the default destination in the navigation.
 */
class WordFragment : Fragment() {
    private val TAG = WordFragment::class.java.simpleName
    private var _binding: FragmentWordBinding? = null
    private var model: WordFragmentViewModel? = null
    private var adapter:WordAdapter? = null
    private var isExpandable = false
    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {

        _binding = FragmentWordBinding.inflate(inflater, container, false)
        return binding.root

    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.btnNext.setOnClickListener {
            findNavController().navigate(R.id.action_WordFragment_to_LessonFragment)
        }

        binding.btnGet.setOnClickListener { model!!.setQuery("新标日初级_03") }
        adapter = WordAdapter(isExpandable)
        val linearLayoutManager= LinearLayoutManager(context)
        linearLayoutManager.orientation = RecyclerView.VERTICAL
        binding.rvWord.layoutManager = linearLayoutManager
        binding.rvWord.setHasFixedSize(true)
        binding.rvWord.adapter = adapter
        model = ViewModelProvider(this).get(WordFragmentViewModel::class.java)

        model!!.wordList.observe(viewLifecycleOwner) { words: List<Word> ->
            if (words != null && words.size != 0) {
                Log.e(TAG, " words.get(0):" + words[0].toString())
            }
            adapter?.submitList(words)

        }

        Log.e(TAG, "wordList:" + model!!.wordList)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}